### PR TITLE
SDIT-4027: ⬆️ Upgrade to spring boot 4.1.0 and workaround spring security bug

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/service/GlobalPrincipalOAuth2AuthorizedClientService.kt [hunk 1]
```diff
@@ -32,12 +32,12 @@ class GlobalPrincipalOAuth2AuthorizedClientService(
   private val authorizedClients: MutableMap<OAuth2AuthorizedClientId, OAuth2AuthorizedClient> =
     ConcurrentHashMap()
 
-  override fun <T : OAuth2AuthorizedClient?> loadAuthorizedClient(
+  override fun <T : OAuth2AuthorizedClient> loadAuthorizedClient(
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/service/GlobalPrincipalOAuth2AuthorizedClientService.kt:35]: nullable changes as a result of spring security upgrade

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsWebClientConfiguration.kt [hunk 2]
```diff
@@ -170,6 +179,7 @@ fun WebClient.Builder.authorisedWebClient(
 
   return baseUrl(url)
     .clientConnector(ReactorClientHttpConnector(proxyAwareHttpClient(timeout)))
+    .filter(ServletRequestResponseNonNullFilterFunction())
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsWebClientConfiguration.kt:181]: add in special filter to workaround issue with request / response being `null`

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsWebClientConfiguration.kt [hunk 3]
```diff
@@ -300,3 +310,21 @@ private fun WebClientReactiveClientCredentialsTokenResponseClient.configureWebCl
...
 }
+
+class ServletRequestResponseNonNullFilterFunction: ExchangeFilterFunction {
+  private companion object {
+    private val HTTP_SERVLET_REQUEST_ATTR_NAME: String = HttpServletRequest::class.java.getName()
+    private val HTTP_SERVLET_RESPONSE_ATTR_NAME: String = HttpServletResponse::class.java.getName()
+  }
+
+  override fun filter(request: ClientRequest, next: ExchangeFunction): Mono<ClientResponse> {
+    val servletRequest = request.attribute(HTTP_SERVLET_REQUEST_ATTR_NAME).getOrNull() ?: RequestFacade(null)
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsWebClientConfiguration.kt:320]: deliberately do nothing if there is already a not null request / response

File: hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsWebClientConfiguration.kt [hunk 4]
```diff
@@ -300,3 +310,21 @@ private fun WebClientReactiveClientCredentialsTokenResponseClient.configureWebCl
       .build(),
   )
 }
+
+class ServletRequestResponseNonNullFilterFunction: ExchangeFilterFunction {
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-autoconfigure/src/main/kotlin/uk/gov/justice/hmpps/kotlin/auth/HmppsWebClientConfiguration.kt]: public so that others can use directly if needed

File: test-app/src/test/kotlin/uk/gov/justice/digital/hmpps/testapp/integration/service/PrisonApiServiceTest.kt [hunk 5]
```diff
@@ -24,4 +30,18 @@ class PrisonApiServiceTest : IntegrationTestBase() {
         .withHeader("Authorization", WireMock.equalTo("Bearer ABCDE")),
     )
   }
+
+  @Test
+  fun `should supply authentication token even when calling service directly`() {
+    prisonApi.stubGetPrisonerLatestBooking("A1234AA")
+
+    assertThat(prisonApiService.getOffenderBooking("A1234AA"))
```
Comment (petergphillips) [inline test-app/src/test/kotlin/uk/gov/justice/digital/hmpps/testapp/integration/service/PrisonApiServiceTest.kt:38]: add in test for `null` request / response

File: build.gradle.kts [hunk 6]
```diff
@@ -11,7 +11,7 @@ plugins {
 
 allprojects {
   group = "uk.gov.justice.service.hmpps"
-  version = "2.5.0"
+  version = "3.0.0-beta"
```
Comment (petergphillips) [inline build.gradle.kts:14]: initially publish as a beta

File: hmpps-kotlin-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/kotlin/auth/service/ReactiveGlobalPrincipalOAuth2AuthorizedClientServiceTest.kt [hunk 7]
```diff
@@ -126,30 +126,14 @@ class ReactiveGlobalPrincipalOAuth2AuthorizedClientServiceTest {
         TEST_AUTHORIZED_CLIENT,
       )
     }
-
-    @Test
-    fun `saveAuthorizedClient stores OAuth2AuthorizedClient under the system username when the authenticated principal is null`() {
```
Comment (petergphillips) [inline hmpps-kotlin-spring-boot-autoconfigure/src/test/kotlin/uk/gov/justice/hmpps/kotlin/auth/service/ReactiveGlobalPrincipalOAuth2AuthorizedClientServiceTest.kt:131]: new spring security null checks ensure that this will never happen

**Review by mikehalmamoj — APPROVED** [review]
